lib/charms/istio_k8s/v0/istio_info.py

```diff
@@ -0,0 +1,239 @@
+
+    def __init__(
+            self,
+            charm: CharmBase,
+            relation_name: str = DEFAULT_RELATION_NAME,
+    ) -> None:
+        super().__init__(charm, relation_name)
+
+        self._charm = charm
+        self._relation_name = relation_name
+
+        self.framework.observe(
+            self._charm.on[self._relation_name].relation_changed,
+            self.on_relation_changed
+        )
+
+    def on_relation_changed(self, _: EventBase) -> None:
+        self.on.info_changed.emit()
+
+    def _get_relations(self):
+        """Return the applications related to us under the monitored relation."""
+        return self._charm.model.relations.get(self._relation_name, tuple())
+
+    def get_data(self, skip_empty: bool = True) -> Optional[IstioInfoAppData]:
+        """Return data for at most one related application, raising if more than one is available.
+
+        Useful for charms that always expect exactly one related application.  It is recommended that those charms also
+        set limit=1 for that relation in charmcraft.yaml.  Returns None if no data is available.
+
+        # TODO: Should this raise if no data is available?  Or just return None?
+
+        Args:
+            skip_empty: If True, return None if there is a relation that has not yet provided any data.
+                        If False, raise a DataValidationError
+                        TODO: Is there a better way to handle the False case?  Practically, if we ever don't skip this
+                         case it'll be hard for a charm to avoid erroring on transient situations while data is coming
+                         but not here yet.
+        """
+        # TODO: Use get_data_from_all_relations here
+        relations = self._get_relations()
+        if len(relations) == 0:
+            return None
+        if len(relations) > 1:
+            # TODO: Different exception type?
+            raise ValueError("Cannot get_info when more than one application is related.")
+
+        raw_data = relations[0].data.get(relations[0].app)
+        if raw_data == {} and skip_empty:
+            return None
+
+        return IstioInfoAppData.load(relations[0].data[relations[0].app])
+
+    def get_data_from_all_relations(self, skip_empty: bool = True) -> List[IstioInfoAppData]:
+        """Return a list of data objects from all relations.
+
+        Args:
+            skip_empty: If True, return None if there is a relation that has not yet provided any data.
+                        If False, raise a DataValidationError
+                        TODO: is it practical for this to ever be used with skip_empty=False?  Should it be removed?
+        """
+        info_list = []
+        for relation in self._get_relations():
+            data = relation.data.get(relation.app)
+            if data == {} and skip_empty:
+                continue
+            info_list.append(IstioInfoAppData.load(relation.data[relation.app]))
+        return info_list
```

Comment:
Feedback is appreciated here.  The goal is for this library to handle both:
* when this relation is set to limit=1 and thus only one data structure is returned
* when this relation can have N related applications and thus can return N items

My thought was that, for the case of limit=1, it is ugly UX if `get_data` returns a list with one element (eg: `[IstioInfoAppData]`), so it makes sense to have two getters (one for limit=1 case that returns a single data structure, and one for all cases that returns the list).  But the current implementation feels messy (name sucks, not sure about skip_empty, etc).

Reply:
Also, unless we come to consensus really fast on this, we can park the discussion and just implement the above `get_data`.  Istio, grafana, tempo, and prometheus all can be captured as limit=1 relations for now, so we can leave the discussion open

Reply:
Yes I agree this approach is the way to go for sure. 

Though ` get_data_from_all_relations` feel a bit verbose.

Reply:
yeah the name feels awful.  If there's agreement that it should eventually be here, then I can include a note about it and when we have a need/better name we can actually implement

Reply:
Discussion moved [here](https://github.com/canonical/charm-relation-building-blocks/pull/1/files#r1941783024)